File: app/invoice_builder.py
import re

from app.models import InvoiceData, LineItem


def _is_blank_row(row: dict) -> bool:
    return not row.get("date", "").strip() and not row.get(
        "client", ""
    ).strip() and not row.get("hours", "").strip()


def validate(form: dict) -> list:
    errors = []

    if not form.get("bill_to_name", "").strip():
        errors.append("Enter the client's name.")

    try:
        hourly_rate = float(form.get("hourly_rate", ""))
        if hourly_rate <= 0:
            errors.append("Hourly rate must be greater than zero.")
    except ValueError:
        errors.append("Invalid hourly rate.")

    rows = [row for row in form.get("items", []) if not _is_blank_row(row)]
    if not rows:
        errors.append("Add at least one item.")

    for index, row in enumerate(rows, start=1):
        if not row.get("client", "").strip():
            errors.append(f"Item {index}: client is required.")

        try:
            hours = float(row.get("hours", ""))
            if hours <= 0:
                errors.append(f"Item {index}: Amount must be greater than zero.")
        except ValueError:
            errors.append(f"Item {index}: Invalid amount.")

    return errors


def build_invoice_data(form: dict) -> InvoiceData:
    items = []
    for row in form.get("items", []):
        if _is_blank_row(row):
            continue
        items.append(
            LineItem(
                date=row["date"].strip(),
                client=row["client"].strip(),
                hours=float(row["hours"]),
            )
        )

    return InvoiceData(
        from_name=form["from_name"],
        from_phone=form["from_phone"],
        from_address=form["from_address"],
        bill_to_name=form["bill_to_name"].strip(),
        bill_to_phone=form["bill_to_phone"],
        bill_to_address=form["bill_to_address"],
        invoice_no=form["invoice_no"],
        date_from=form["date_from"],
        date_to=form["date_to"],
        week_from=form["week_from"],
        week_to=form["week_to"],
        payment_method=form["payment_method"],
        hourly_rate=float(form["hourly_rate"]),
        items=items,
    )


def suggested_filename(invoice_no: str, bill_to_name: str, ext: str) -> str:
    safe_name = re.sub(r"[^A-Za-z0-9]+", "_", bill_to_name).strip("_")
    return f"Invoice_{invoice_no}_{safe_name}.{ext}"

File: webapp/js/invoiceBuilder.js
(function () {
  const SLV = (globalThis.SLV = globalThis.SLV || {});
  const { LineItem, InvoiceData } = SLV.models;

  const text = (value) => (value === undefined || value === null ? "" : String(value));

  /** Parses like Python's float(): blank and non-numeric input are rejected. */
  function parseNumber(value) {
    const trimmed = text(value).trim();
    if (!trimmed) return NaN;
    return Number(trimmed);
  }

  function isBlankRow(row) {
    return (
      !text(row.date).trim() && !text(row.client).trim() && !text(row.hours).trim()
    );
  }

  function validate(form) {
    const errors = [];

    if (!text(form.billToName).trim()) {
      errors.push("Enter the client's name.");
    }

    const hourlyRate = parseNumber(form.hourlyRate);
    if (Number.isNaN(hourlyRate)) {
      errors.push("Invalid hourly rate.");
    } else if (hourlyRate <= 0) {
      errors.push("Hourly rate must be greater than zero.");
    }

    const rows = (form.items || []).filter((row) => !isBlankRow(row));
    if (rows.length === 0) {
      errors.push("Add at least one item.");
    }

    rows.forEach((row, index) => {
      const position = index + 1;

      if (!text(row.client).trim()) {
        errors.push(`Item ${position}: client is required.`);
      }

      const hours = parseNumber(row.hours);
      if (Number.isNaN(hours)) {
        errors.push(`Item ${position}: Invalid amount.`);
      } else if (hours <= 0) {
        errors.push(`Item ${position}: Amount must be greater than zero.`);
      }
    });

    return errors;
  }

  function buildInvoiceData(form) {
    const items = (form.items || [])
      .filter((row) => !isBlankRow(row))
      .map(
        (row) =>
          new LineItem({
            date: text(row.date).trim(),
            client: text(row.client).trim(),
            hours: parseNumber(row.hours),
          })
      );

    return new InvoiceData({
      fromName: text(form.fromName),
      fromPhone: text(form.fromPhone),
      fromAddress: text(form.fromAddress),
      billToName: text(form.billToName).trim(),
      billToPhone: text(form.billToPhone),
      billToAddress: text(form.billToAddress),
      invoiceNo: text(form.invoiceNo),
      dateFrom: text(form.dateFrom),
      dateTo: text(form.dateTo),
      weekFrom: text(form.weekFrom),
      weekTo: text(form.weekTo),
      paymentMethod: text(form.paymentMethod),
      hourlyRate: parseNumber(form.hourlyRate),
      items,
    });
  }

  function suggestedFilename(invoiceNo, billToName, ext) {
    const safeName = text(billToName).replace(/[^A-Za-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
    return `Invoice_${invoiceNo}_${safeName}.${ext}`;
  }

  SLV.invoiceBuilder = { validate, buildInvoiceData, suggestedFilename };
})();
